### PR TITLE
faster calculation of QGramDistances if precounting qgrams either as …

### PR DESCRIPTION
**Pull request by robertfeldt**
…Dict or sorted array, with tests

This is the PR based on my gists showing how to speed up QGramDistance calculation in scenarios when you can precount/precalculate the QGram counts.

I'm seeing 10x-30x speedups on my machine/setup depending on string length and length of qgrams (q). Also memory use is much reduced. Example code:

````julia
julia> using StringDistances, Random, BenchmarkTools

julia> s1, s2 = randstring(1000), randstring(800);

julia> p1, p2 = QGramSortedArray(s1, 2), QGramSortedArray(s2, 2);

julia> @btime evaluate(Jaccard(2), s1, s2)
  315.392 μs (1811 allocations: 495.89 KiB)
0.8906574394463668

julia> @btime evaluate(Jaccard(2), p1, p2)
  11.788 μs (2 allocations: 48 bytes)
0.8906574394463668

julia> speedup = 315.392 / 11.788
26.755344418052257
```

### CODE REVIEW

File: src/distances/qgram.jl [hunk 1]
```diff
@@ -80,9 +78,141 @@ function _count(s1, s2)
...
+    counts::Dict{K,Int}
+end
+function QGramDict(s::Union{AbstractString, AbstractVector}, q::Integer = 2)
+    @assert q >= 1
+    qgs = qgrams(s, q)
+    QGramDict{q, eltype(qgs)}(countdict(qgs))
+end
+
+# Faster (than QgramDict) with the qgrams presorted
+struct QGramSortedArray{Q,K} <: AbstractQGramCounts{Q,K}
```
Comment (matthieugomez) [inline src/distances/qgram.jl]: QGramSortedVector sounds like a better name?

File: src/distances/qgram.jl [hunk 2]
```diff
@@ -80,9 +78,141 @@ function _count(s1, s2)
...
+		end
+    end
+    d
+end
+
+abstract type AbstractQGramCounts{Q,K} end
+q(qc::AbstractQGramCounts{Q,K}) where {Q,K} = Q
+counts(qc::AbstractQGramCounts) = qc.counts
+
+struct QGramDict{Q,K} <: AbstractQGramCounts{Q,K}
```
Comment (matthieugomez) [inline src/distances/qgram.jl:120]: Do you really want to have both QGramDict and QGramSortedArray? Would not it be simpler to only have QGramSortedArray?

File: test/distances.jl [hunk 3]
```diff
@@ -130,8 +123,31 @@ using StringDistances, Unicode, Test
...
+				qd2 = QGramDict(str2, qlen)
+				expected = evaluate(dist, str1, str2)
+				@test expected == evaluate(dist, qd1, qd2)
+
+				# QGramSortedArray gets same result as for standard string
+				qc1 = QGramSortedArray(str1, qlen)
+				qc2 = QGramSortedArray(str2, qlen)
+				@test expected == evaluate(dist, qc1, qc2)
+			end
+		end
```
Comment (matthieugomez) [inline test/distances.jl:217]: Could you add tests for Unicode characters, as well as arbitrary iterators (i.e. vectors or `graphemes`)?

**Review by matthieugomez — COMMENTED** [review]
Awesome! I just left three comments. Thanks a lot for your help!

File: src/distances/qgram.jl [hunk 4]
```diff
@@ -80,9 +78,141 @@ function _count(s1, s2)
...
+    @assert q >= 1
+    qgs = qgrams(s, q)
+    QGramDict{q, eltype(qgs)}(countdict(qgs))
+end
+
+# Faster (than QgramDict) with the qgrams presorted
+struct QGramSortedArray{Q,K} <: AbstractQGramCounts{Q,K}
+    counts::Vector{Pair{K,Int}}
+end
+function QGramSortedArray(s::Union{AbstractString, AbstractVector}, q::Integer = 2)
```
Comment (matthieugomez) [inline src/distances/qgram.jl]: Could it work with arbitrary iterator that define `length` (like `graphemes` for instance)?

Comment (robertfeldt) [reply src/distances/qgram.jl]: I would think so. I'm not very familiar with Unicode or `graphemes` but if your existing functionality supports it I should be able to add it. Will check. Would fulfilling tests like this be what you are after:

````julia
s1, s2 = "alborgów", "amoniak"
p1, p2 = QGramSortedVector(s1), QGramSortedVector(s2)
p1b, p2b = QGramSortedVector(graphemes(s1)), QGramSortedVector(graphemes(s2))
@test evaluate(Jaccard(2), s1, s2) == evaluate(Jaccard(2), p1, p2)
@test evaluate(Jaccard(2), graphemes(s1), graphemes(s2)) == evaluate(Jaccard(2), p1, p2)
@test evaluate(Jaccard(2), graphemes(s1), graphemes(s2)) == evaluate(Jaccard(2), p1b, p2b)
````

File: test/distances.jl (hunk 3 above)
Comment (robertfeldt) [reply test/distances.jl:217]: I think so but please see my question on this below...

**Comment by robertfeldt** [comment]
This seems to only fail on older Julia versions since we then couldn't add methods on an abstract type: `LoadError: cannot add methods to an abstract type`. Fixing this will be quite cumbersome, I think, since we would need to basically `eval` the same method implementation on all of the concrete sub-types. Do we really need to support very old Julia versions?

File: src/distances/qgram.jl (hunk 2 above)
Comment (robertfeldt) [reply src/distances/qgram.jl:120]: Yes, I would very strongly prefer to keep both in there. At least two benefits to the QGramDict one:
1. We can support streaming QGram counting (think very large strings/streams),
2. We can do set "arithmetics" on QGramDicts for specific distances or when adding Diversity/Diameter distances that work on sets/vectors of strings.

I would argue that the benefits outweigh the downside of having to support also QGramDict. The code for it is limited and simple so shouldn't be a problem.

If anything, I would propose extracting the iterators and QGram dict/sortedvector code to a qgram_helpers.jl or qgram_counting.jl file to keep things more modular and separate. This latter file would be more nitty-gritty and not needed for people that just want to look at the QGramDistances and the code that implements them.

File: src/distances/qgram.jl (hunk 1 above)
Comment (robertfeldt) [reply src/distances/qgram.jl]: Agreed, will fix.

**Comment by robertfeldt** [comment]
From Julia 1.3 and on it has been possible to add methods on an abstract type:

````julia
feldt:~/tmp$ cat method_on_abstract_type.jl 
@show VERSION

abstract type A end

struct C <: A
    c::Int
end
counter(c::C) = c.c

c = C(19)

(a::A)(x::Int) = x + counter(a)

@show c(1)
feldt:~/tmp$ julia15 method_on_abstract_type.jl 
VERSION = v"1.5.2"
c(1) = 20
feldt:~/tmp$ julia14 method_on_abstract_type.jl 
VERSION = v"1.4.2"
c(1) = 20
feldt:~/tmp$ julia13 method_on_abstract_type.jl 
VERSION = v"1.3.1"
c(1) = 20
feldt:~/tmp$ julia12 method_on_abstract_type.jl 
VERSION = v"1.2.0"
ERROR: LoadError: cannot add methods to an abstract type
Stacktrace:
 [1] top-level scope at /Users/feldt/tmp/method_on_abstract_type.jl:12
 [2] include at ./boot.jl:328 [inlined]
 [3] include_relative(::Module, ::String) at ./loading.jl:1094
 [4] include(::Module, ::String) at ./Base.jl:31
 [5] exec_options(::Base.JLOptions) at ./client.jl:295
 [6] _start() at ./client.jl:464
in expression starting at /Users/feldt/tmp/method_on_abstract_type.jl:12
````

Can we require Julia 1.3 or do you have some nice solution to this?

**Comment by robertfeldt** [comment]
Ok, I think I fixed most of your comments now. Quite some more tests. Feedback appreciated. Main things:

1. Did the name change to QGramSortedVector
2. Ensured can handle Unicode strings and graphemes output
3. Tests for Unicode etc

**Comment by matthieugomez** [comment]
Thanks for answering my comments. Last thing: can you add some `"""... """` doc for `QGramSortedVector` and `QgramDict` since they are exported?

It is a matter of taste, but I'm not a big fan of how the code is split into `calculate` `countleft` etc, I think it makes it hard to get a high-level view of what happens when I call `QGram(2)("ok", "ok")`. I will merge it as is, but I may tinker a bit with it once it's merged — I hope that's ok with you.

**Comment by matthieugomez** [comment]
I don't mind requiring Julia 1.3 but then you need to change the `.travis.yml` file and the `Project.toml` file.

**Comment by robertfeldt** [comment]
Ok, will look into documentation tomorrow.

I agree it makes it harder to understand but I found it essential for best performance and I do think that is a priority. If you see a way to keep performance while making it more "collected" I'm all for it. But this division allows the `countmatches!` to specialise a lot for individual types. But yeah, I'm only attached to the performance not to the design itself so if you find a better one that is perfectly fine. Please tell me though since then I can move things back to my MultiDist.jl package. Cheers.

**Comment by matthieugomez** [comment]
Yes I will merge it as is, and only change things if they do not affect performances. In any case, your pull request is very welcome: it's great to have a much faster way to compute distances.

**Comment by robertfeldt** [comment]
Ok, doc strings added and upped Julia to 1.3. Hopefully, this should build and test fine.

Are you also interested in a PR for pairwise and for more compression distances? I'd rather collect valuable code into one powerful Julia package than spread it out over multiple packages.

Again, thanks for the package, we us string distances a lot in our research so this is going to be very useful for software testing packages and other things coming up over time.

Cheers!

**Comment by robertfeldt** [comment]
Hmm, we get a failed test since NaN != NaN. But I don't have the strings that caused this. Not sure if this would happen also before this PR. We could just change the test assertion but maybe we should never get a NaN so there is a deeper issue here. Will investigate.

**Comment by robertfeldt** [comment]
I re-ran tests several times on both 1.5 and 1.3 but cannot replicate this. Can it be a Linux thing? Probably not, and just unlikely and due to the random tests. Hmm.

**Comment by matthieugomez** [comment]
Thanks again for the pull request. Yes, I am interested in more pull requests. I don't really use string distances in my own research so you're welcome to shape the package more towards your needs.

For the failed test, is it because you use random strings in your tests? Could you remove that (e.g. hardcode a particular set of strings), so that we don't have random errors?

**Comment by robertfeldt** [comment]
Yes, very likely due to random testing. OTOH, it is not a behaviour we ever want to see. I'll see if I can replicate.

**Comment by matthieugomez** [comment]
Yes, I agree we need to correct it. I'm just saying I don't like random tests, since then it's not clear whether a PR fails because of new code, or because of different random sample. So I'd prefer to have a hardcoded set of strings.

**Comment by robertfeldt** [comment]
Agreed, will fix (if real bug) or revert to fixed test strings.

**Comment by robertfeldt** [comment]
I found it (and without random testing maybe I wouldn't have). This happens if the q is larger than the length of the string. How do you want us to handle this? How was it handled before?

**Comment by robertfeldt** [comment]
I had to up the number of random tests a lot for this to surface on my machine so the Travis CI run was a bit lucky/unlucky... ;)

**Comment by matthieugomez** [comment]
I see, thanks. Yes it did give NaN before. I think it makes sense, but a
DomainError would also makes sense.

**Comment by robertfeldt** [comment]
This PR should fix the test issue with short strings:

https://github.com/matthieugomez/StringDistances.jl/pull/37
